lib/Service/Classification/ImportanceClassifier.php

```diff
@@ -0,0 +1,311 @@
+use function array_map;
+use function array_slice;
+
+class ImportanceClassifier {
+
+	/**
+	 * Mailbox special uses to exclude from the training
+	 */
+	private const EXEMPT_FROM_TRAINING = [
+		Horde_Imap_Client::SPECIALUSE_ALL,
+		Horde_Imap_Client::SPECIALUSE_DRAFTS,
+		Horde_Imap_Client::SPECIALUSE_FLAGGED,
+		Horde_Imap_Client::SPECIALUSE_JUNK,
+		Horde_Imap_Client::SPECIALUSE_SENT,
+		Horde_Imap_Client::SPECIALUSE_TRASH,
+	];
+
+	/**
+	 * @var string label for data sets that are classified as important
+	 */
+	private const LABEL_IMPORTANT = 'i';
+
+	/**
+	 * @var string label for data sets that are classified as not important
+	 */
+	private const LABEL_NOT_IMPORTANT = 'ni';
+
+	/**
+	 * The maximum number of data sets to train the classifier with
+	 */
+	private const MAX_TRAINING_SET_SIZE = 1000;
+
+	/** @var MailboxMapper */
+	private $mailboxMapper;
+
+	/** @var MessageMapper */
+	private $messageMapper;
+
+	/** @var CompositeExtractor */
+	private $extractor;
+
+	/** @var PersistenceService */
+	private $persistenceService;
+
+	/** @var PerformanceLogger */
+	private $performanceLogger;
+
+	/** @var ILogger */
+	private $logger;
+
+	public function __construct(MailboxMapper $mailboxMapper,
+								MessageMapper $messageMapper,
+								CompositeExtractor $extractor,
+								PersistenceService $persistenceService,
+								PerformanceLogger $performanceLogger,
+								ILogger $logger) {
+		$this->mailboxMapper = $mailboxMapper;
+		$this->messageMapper = $messageMapper;
+		$this->extractor = $extractor;
+		$this->persistenceService = $persistenceService;
+		$this->performanceLogger = $performanceLogger;
+		$this->logger = $logger;
+	}
+
+	/**
+	 * Train an account's classifier of important messages
+	 *
+	 * Train a classifier based on a user's existing messages to be able to derive
+	 * importance markers for new incoming messages.
+	 *
+	 * To factor in (server-side) filtering into multiple mailboxes, the algorithm
+	 * will not only look for messages in the inbox but also other non-special
+	 * mailboxes.
+	 *
+	 * To prevent memory exhaustion, the process will only load a fixed maximum
+	 * number of messages per account.
+	 *
+	 * @param Account $account
+	 */
+	public function train(Account $account): void {
+		$perf = $this->performanceLogger->start('importance classifier training');
+		$incomingMailboxes = $this->getIncomingMailboxes($account);
+		$perf->step('find incoming mailboxes');
+		$outgoingMailboxes = $this->getOutgoingMailboxes($account);
+		$perf->step('find outgoing mailboxes');
+
+		$mailboxIds = array_map(function (Mailbox $mailbox) {
+			return $mailbox->getId();
+		}, $incomingMailboxes);
+		$messages = array_filter(
+			$this->messageMapper->findLatestMessages($mailboxIds, self::MAX_TRAINING_SET_SIZE),
+			function(Message $message) {
+				return $message->getFrom() !== null;
+			}
+		);
+		$importantMessages = array_filter($messages, function(Message $message) {
+			return $message->getFlagImportant();
```

Comment:
```suggestion
			return $message->getFlagImportant() || $message->getFlagFlagged();
```

I guess if one flagged a message it's already important? ;)

Reply:
No, that should be separate actually :)

And also here we just make sure that there are important messages so the training makes sense.